### ISSUE TITLE
Assessment job not started due to error: Microsoft Graph service exception

### ISSUE THREAD

**Issue by Joe-Kn** [pnp/pnpassessment#47]
Hello,

I am sorry but I am totally new to M365 Assessment and just followed the instructions. I created the Azure App and granted the permissions:

![permissionsgranted](https://github.com/pnp/pnpassessment/assets/164862351/0462fd35-4ca5-487d-bf65-0f18e79396c5)

However, I am getting an error that the assessment job could not be started. I need an assessment for all add ins.

This is what I see at the logs:

[ERR] Code in PreScanningAsync failed. This can happen when the first enumerated site collection is locked
HttpResponseCode: 403
Code: Authorization_RequestDenied
Message: Insufficient privileges to complete the operation.
PnP.Core.MicrosoftGraphServiceException: Microsoft Graph service exception
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphInteractiveAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphBatchAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
   at PnP.Core.Model.BaseDataModel`1.RequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Model.BaseDataModel`1.RawRequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Admin.Model.SharePoint.LegacyPrincipalManagement.GetValidLegacyServicePrincipalAppIdsAsync(PnPContext context, Boolean includeExpiredPrincipals)
   at PnP.Core.Admin.Model.SharePoint.SiteCollectionManager.GetLegacyServicePrincipalsAsync(Boolean includeExpiredPrincipals)
   at PnP.Scanning.Core.Scanners.AddInACSScanComponent.ExecutePreScanningAsync(ScannerBase scannerBase, PnPContext context, ClientContext csomContext, VanityUrlOptions vanityUrlOptions)
   at PnP.Scanning.Core.Scanners.AddInACSScanner.PreScanningAsync()
   
   [ERR] Preassessment for assessment "023c2da1-f47b-49cd-bf5d-e6754b31cf87" failed. Error: Microsoft Graph service exception
HttpResponseCode: 403
Code: Authorization_RequestDenied
Message: Insufficient privileges to complete the operation.
PnP.Core.MicrosoftGraphServiceException: Microsoft Graph service exception
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphInteractiveAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphBatchAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
   at PnP.Core.Model.BaseDataModel`1.RequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Model.BaseDataModel`1.RawRequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Admin.Model.SharePoint.LegacyPrincipalManagement.GetValidLegacyServicePrincipalAppIdsAsync(PnPContext context, Boolean includeExpiredPrincipals)
   at PnP.Core.Admin.Model.SharePoint.SiteCollectionManager.GetLegacyServicePrincipalsAsync(Boolean includeExpiredPrincipals)
   at PnP.Scanning.Core.Scanners.AddInACSScanComponent.ExecutePreScanningAsync(ScannerBase scannerBase, PnPContext context, ClientContext csomContext, VanityUrlOptions vanityUrlOptions)
   at PnP.Scanning.Core.Scanners.AddInACSScanner.PreScanningAsync()
   at PnP.Scanning.Core.Services.ScanManager.StartScanAsync(StartRequest start, AuthenticationManager authenticationManager, List`1 siteCollectionList)
   
   Which permissions are required or wrong?
   
   I used the "Start a new SharePoint Add-In and Azure ACS assessment (application permissions) for a complete tenant" command. 
   
   You are using the latest version 1.7.0

Connecting Microsoft 365 Assessment on port 25010...
OK

Initializing authentication
Starting the Microsoft 365 Assessment
Microsoft 365 Assessment authentication initialized
Enumerated 8720 site collections for tenant <tenant>.sharepoint.com
Sites to assess are defined
Assessment job not started due to error: Microsoft Graph service exception
   
   Many thanks for help
   Joe
   
   PS: I added my account I use for the assessment to all Site Collections within our tenant as a SCA.

**Comment by Joe-Kn** [comment]
This is what the second log tells:

2024-03-25 09:46:51.327 +01:00 [ERR] Code in PreScanningAsync failed. This can happen when the first enumerated site collection is locked
HttpResponseCode: 403
Code: Authorization_RequestDenied
Message: Insufficient privileges to complete the operation.
ClientRequestId: ce39ef7a-f2b2-431c-a0e2-659685083d30
date: 2024-03-25T08:46:53
client-request-id: ce39ef7a-f2b2-431c-a0e2-659685083d30

PnP.Core.MicrosoftGraphServiceException: Microsoft Graph service exception
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphInteractiveAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphBatchAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
   at PnP.Core.Model.BaseDataModel`1.RequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Model.BaseDataModel`1.RawRequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Admin.Model.SharePoint.LegacyPrincipalManagement.GetValidLegacyServicePrincipalAppIdsAsync(PnPContext context, Boolean includeExpiredPrincipals)
   at PnP.Core.Admin.Model.SharePoint.SiteCollectionManager.GetLegacyServicePrincipalsAsync(Boolean includeExpiredPrincipals)
   at PnP.Scanning.Core.Scanners.AddInACSScanComponent.ExecutePreScanningAsync(ScannerBase scannerBase, PnPContext context, ClientContext csomContext, VanityUrlOptions vanityUrlOptions)
   at PnP.Scanning.Core.Scanners.AddInACSScanner.PreScanningAsync()
2024-03-25 09:46:51.348 +01:00 [ERR] Preassessment for assessment "023c2da1-f47b-49cd-bf5d-e6754b31cf87" failed. Error: Microsoft Graph service exception
HttpResponseCode: 403
Code: Authorization_RequestDenied
Message: Insufficient privileges to complete the operation.
ClientRequestId: ce39ef7a-f2b2-431c-a0e2-659685083d30
date: 2024-03-25T08:46:53
client-request-id: ce39ef7a-f2b2-431c-a0e2-659685083d30

PnP.Core.MicrosoftGraphServiceException: Microsoft Graph service exception
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphInteractiveAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphBatchAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
   at PnP.Core.Model.BaseDataModel`1.RequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Model.BaseDataModel`1.RawRequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Admin.Model.SharePoint.LegacyPrincipalManagement.GetValidLegacyServicePrincipalAppIdsAsync(PnPContext context, Boolean includeExpiredPrincipals)
   at PnP.Core.Admin.Model.SharePoint.SiteCollectionManager.GetLegacyServicePrincipalsAsync(Boolean includeExpiredPrincipals)
   at PnP.Scanning.Core.Scanners.AddInACSScanComponent.ExecutePreScanningAsync(ScannerBase scannerBase, PnPContext context, ClientContext csomContext, VanityUrlOptions vanityUrlOptions)
   at PnP.Scanning.Core.Scanners.AddInACSScanner.PreScanningAsync()
   at PnP.Scanning.Core.Services.ScanManager.StartScanAsync(StartRequest start, AuthenticationManager authenticationManager, List`1 siteCollectionList)
2024-03-25 09:46:51.351 +01:00 [INF] Setting Scan table to preassessmentstatus "Failed" for assessment "023c2da1-f47b-49cd-bf5d-e6754b31cf87"
2024-03-25 09:46:51.361 +01:00 [INF] Database updates pushed in SetPreScanStatusAsync for assessment "023c2da1-f47b-49cd-bf5d-e6754b31cf87"
2024-03-25 09:46:51.363 +01:00 [ERR] Error starting assessment job: Microsoft Graph service exception
HttpResponseCode: 403
Code: Authorization_RequestDenied
Message: Insufficient privileges to complete the operation.
ClientRequestId: ce39ef7a-f2b2-431c-a0e2-659685083d30
date: 2024-03-25T08:46:53
client-request-id: ce39ef7a-f2b2-431c-a0e2-659685083d30

PnP.Core.MicrosoftGraphServiceException: Microsoft Graph service exception
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphInteractiveAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteMicrosoftGraphBatchAsync(Batch batch)
   at PnP.Core.Services.BatchClient.ExecuteBatch(Batch batch)
   at PnP.Core.Model.BaseDataModel`1.RequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Model.BaseDataModel`1.RawRequestAsync(ApiCall apiCall, HttpMethod method, String operationName)
   at PnP.Core.Admin.Model.SharePoint.LegacyPrincipalManagement.GetValidLegacyServicePrincipalAppIdsAsync(PnPContext context, Boolean includeExpiredPrincipals)
   at PnP.Core.Admin.Model.SharePoint.SiteCollectionManager.GetLegacyServicePrincipalsAsync(Boolean includeExpiredPrincipals)
   at PnP.Scanning.Core.Scanners.AddInACSScanComponent.ExecutePreScanningAsync(ScannerBase scannerBase, PnPContext context, ClientContext csomContext, VanityUrlOptions vanityUrlOptions)
   at PnP.Scanning.Core.Scanners.AddInACSScanner.PreScanningAsync()
   at PnP.Scanning.Core.Services.ScanManager.StartScanAsync(StartRequest start, AuthenticationManager authenticationManager, List`1 siteCollectionList)
   at PnP.Scanning.Core.Services.ScanManager.StartScanAsync(StartRequest start, AuthenticationManager authenticationManager, List`1 siteCollectionList)
   at PnP.Scanning.Core.Services.Scanner.Start(StartRequest request, IServerStreamWriter`1 responseStream, ServerCallContext context)

**Comment by jansenbe** [comment]
@Joe-Kn : can you share the cmdline you've used? When using application permissions the Azure AD app also requires a certificate, see https://pnp.github.io/pnpassessment/using-the-assessment-tool/setupauth.html#creating-an-azure-ad-application-using-the-azure-ad-portal for the details on the Azure AD app configuration.

Next to that, when running an assessment for AddInsACS you also need to specify the application.read.all permission (see https://pnp.github.io/pnpassessment/addinsacs/requirements.html). Given that the Graph calls are failing I think that's the root cause of your issue.

**Comment by Joe-Kn** [comment]
Hello @jansenbe,

many thanks I was now able to start the assessment. 

First of all, it really was the missing "Application.Read.All" permission. The reason I missed this, is that the webpage "[Add-Ins and ACS](https://pnp.github.io/pnpassessment/addinsacs/readme.html)" was having an example command line on how to add the Azure Application and I just re-used the command line, not checking the requirements page on top.

So I just used:

Register-PnPAzureADApp -ApplicationName Microsoft365AssessmentToolForSharePoint -Tenant tenant.onmicrosoft.com -Store CurrentUser -GraphApplicationPermissions "Sites.Read.All" -SharePointApplicationPermissions "Sites.FullControl.All" -GraphDelegatePermissions "Sites.Read.All", "User.Read" -SharePointDelegatePermissions "AllSites.Manage" -Username "my@email.here" -Interactive

but it should have been:

Register-PnPAzureADApp -ApplicationName Microsoft365AssessmentToolForSharePoint -Tenant tenant.onmicrosoft.com -Store CurrentUser -GraphApplicationPermissions "Sites.Read.All", **"Application.Read.All"** -SharePointApplicationPermissions "Sites.FullControl.All" -GraphDelegatePermissions "Sites.Read.All", "User.Read", **"Application.Read.All"** -SharePointDelegatePermissions "AllSites.Manage" -Username "my@email.here" -Interactive

and to start the assessment I used also the example with my parameters of course:

.\microsoft365-assessment.exe start --mode AddInsACS --authmode application --tenant mytenant.sharepoint.com --applicationid 5869cfff-XXX-XXX-XXXX-6cc4cdb8e581 --certpath "My|CurrentUser|B3BXXXXXXXXXXXXXXXXXXXXXXX8960ABEA"


Now the status shows it's running:

Live running Microsoft 365 Assessment status. Press ESC to exit


│                  Id                  │   Mode    │ Status  │    Progress     │ Retries │    Session start     │ Session duration │
│ 6fed4d96-963a-4d13-8949-2b2843f2f7e1 │ AddInsACS │ Running │ 1070/8732 (12%) │  0 / 0  │ 3/26/2024 6:15:22 AM │   00:00:11:01    │


Again many thanks.

Have a great day Bert!
